File: Python/Binary Tree Preorder Traversal .py
"""
Given a binary tree, return the preorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [1,2,3].

Note: Recursive solution is trivial, could you do it iteratively?
"""

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        nodes = []
        result = []
        
        if root== None:
            return result
        nodes.append(root)
        while nodes:
            node = nodes.pop()

            result.append(node.val)
            if node.right:
                nodes.append(node.right)
            if node.left:
                nodes.append(node.left)

        return result

File: Java/src/Binary Tree Preorder Traversal.java
import java.util.LinkedList;

import javax.swing.tree.TreeNode;

/**
 * Definition for binary tree
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */
public class Solution
{
  public List<Integer> preorderTraversal(TreeNode root)
  {
    List<Integer> result = new LinkedList<>();
    LinkedList<TreeNode> nodes = new LinkedList<>();

    if (root == null)
      return result;
    nodes.addLast(root);
    TreeNode node;
    while (nodes.size() > 0)
    {
      node = nodes.removeLast();
      result.add(node.val);
      if (node.right != null)
        nodes.addLast(node.right);
      if (node.left != null)
        nodes.addLast(node.left);
    }
    return result;
  }
}
